src/dataloader.py

```diff
@@ -0,0 +1,140 @@
+            "per_device_train_batch_size": self.batch_size["train"],
+            "per_device_eval_batch_size": self.batch_size["validation"],
+            "data_seed": self.seed,
+        }
+
+    @property
+    def trainer_params(self):
+        """
+        Dataset-specific **kwargs to init HF Trainer
+        Ref: https://huggingface.co/docs/transformers/main_classes/trainer#transformers.Trainer
+        """
+        # we initialise different data collators based on causal or masked language modeling
+        if self.mlm_probability:
+            # in MLM we usually only have [CLS] and [SEP] tokens
+            collate_fn = DataCollatorForLanguageModeling(self.tokenizer, mlm=True, mlm_probability=self.mlm_probability)
+        else:
+            # our method of CLM pretraining we chunk the input (see group function) for pretraining with large amounts
+            # of data, hence no [PAD] token. This means the outputs of the dataloader are all `max_seq_len` long.
+            # So here we explicitly use the default data collator which doesn't include any padding, outputting samples as is.
+            collate_fn = DefaultDataCollator(return_tensors="pt")
+        return {
+            "train_dataset": self.datasets["train"],
+            "eval_dataset": self.datasets["validation"],
+            "data_collator": collate_fn,
+        }
+
+    def dataloader(self, split: Literal["train", "validation"] = "train"):
+        torch.manual_seed(self.seed)
+
+        return DataLoader(
+            self.datasets[split],
+            batch_size=self.batch_size[split],
+            num_workers=self.num_workers,
+            collate_fn=DataCollatorForLanguageModeling(
+                self.tokenizer, mlm=(self.mlm_probability != 0), mlm_probability=self.mlm_probability
+            ),
+        )
+
+    def group(self, example_batch):
+        """Concatenate the dataset into chunks of max_seq_length. Don't waste tokens exceeding seqlen."""
+        # Flatten batch
+        concatenated_examples = {k: list(itertools.chain(*example_batch[k])) for k in example_batch.keys()}
+        total_length = len(concatenated_examples[list(example_batch.keys())[0]])
+        # We drop the small remainder, we could add padding if the model supported it instead of
+        # this drop, you can customize this part to your needs.
+        if total_length >= self.max_seq_length:
+            total_length = (total_length // self.max_seq_length) * self.max_seq_length
+        # Split by chunks of max_len.
+        result = {
+            k: [t[i : i + self.max_seq_length] for i in range(0, total_length, self.max_seq_length)]
+            for k, t in concatenated_examples.items()
+        }
+        return result
+
+    def encode(self, example_batch):
+        # tokenize the text
+        features = self.tokenizer(
+            example_batch["text"],
+            # TODO: this line is trying to pad but there is no guarantee that we actually have a PAD token instantiated
```

Comment:
@photomz Yeah this is the padding issue I mentioned. We have to create a unit test for this per your comments. Just double check in the PR. I'll write the unit test later.